_posts/2024-05-11-coroutines-and-deadlocks.md

```diff
@@ -0,0 +1,231 @@
+A key feature of Project Loom are Virtual Threads, which became production-ready with JDK 21. 
+To appreciate the impact of Virtual Threads, let's first review the traditional threading model in Java:
+
+Java Platform threads are mapped 1:1 to OS threads, making them expensive objects. 
+Kotlin coroutines optimize thread usage by being lightweight and non-blocking, requiring far fewer Platform Threads. 
+However, the challenge persists with blocking code, where we offload these tasks to a large number of threads, hoping there are enough threads to handle the load efficiently.
+
+This is where Virtual Threads come into play. Virtual Threads are also very lightweight. 
+They are executed by Carrier Threads, which are mapped 1:1 to OS threads. 
+The key advantage of Virtual Threads is that they are non-blocking. 
+The Java developers retrofitted the `java.io` and `java.util.concurrent` packages so that whenever a Virtual Thread encounters blocking code, 
+it is automatically suspended, and the Carrier Thread is released. 
+This is a game changer in the Java world because it allows legacy blocking applications to become non-blocking with minimal adaptation.
+
+However, there are still some constraints: you cannot use the `synchronized` block with Virtual Threads, 
+as it will still block the Virtual Thread. Therefore, it's important to review your code and any third-party libraries for compatibility.
+
+Most well-known Java libraries are now Virtual Thread-ready. 
+Spring-Boot delivered Virtual Thread support with the 3.2 release. 
+
+The PostgreSQL Driver has been updated to support Virtual Threads a while ago. 
+This allows us to offload previously blocking database access to Virtual Threads, 
+transforming these operations into non-blocking tasks and resolving the deadlock issue without changes to the application logic.
+
+
+```
+val VIRTUAL_THREAD_DISPATCHER = Executors.newVirtualThreadPerTaskExecutor().asCoroutineDispatcher()
+
+val count = newSuspendedTransaction(VIRTUAL_THREAD_DISPATCHER) {
+    FooTable.selectAll().count()
+}
+```
+
+That's it. By introducing the Virtual Thread Dispatcher context and applying it, we transformed our blocking code into responsive, non-blocking code.
+For a detailed example that illustrates both the problem and the solution, check out our project on GitHub: [Example Project](https://github.com/sw-code/swcode-samples/tree/main/virtual-threads/deadlocks).
+
+# Conclusion
+
+Navigating the complexities of integrating Exposed with Kotlin coroutines revealed significant challenges, particularly around deadlocks and blocking operations. 
+Traditional solutions, like increasing database connections, proved costly and inefficient. 
+However, the introduction of Java Virtual Threads from Project Loom offers a transformative solution. 
+Virtual Threads allow us to handle concurrency efficiently, eliminating deadlocks and making our application fully non-blocking without significant changes to our logic.
+
```

Comment:
but deadlocks are still an annoying and ever-present problem, which is not fully solved by virtual threads, right?